### PR TITLE
feat: Wave 1 — search-first skill, confidence thresholds, version manifest

### LINKED ISSUE #91: Version manifest for upgrade tracking
## Problem

No version tracking for installed plugins. Users running `devflow init` can't tell if they're upgrading, reinstalling, or doing a fresh install. No mechanism to trigger data migrations when versions change.

## Proposed Fix

Add `.devflow/manifest.json` with structure:

```json
{
  "version": "1.0.0",
  "plugins": [
    { "name": "devflow-implement", "version": "1.1.0" }
  ],
  "installedAt": "2026-03-07T00:00:00Z",
  "updatedAt": "2026-03-07T00:00:00Z"
}
```

- Check on `devflow init` — compare installed vs package version
- Trigger migrations when version changes (e.g., path renames, setting schema updates)
- Display version info in `devflow list`

## Scope

- Low-medium effort — new file, init-time check, migration framework

### LINKED ISSUE #111: Search-First Skill: Research Before Building
## Summary

Add a `search-first` skill that enforces researching existing solutions before writing custom code. A 4-phase loop: Need Analysis → Parallel Search → Evaluation → Decision. Inspired by Harness Alpha's search-first workflow pattern.

## Motivation

Harness Alpha identifies a common failure mode in AI-assisted development: **building custom utilities when battle-tested packages already exist**. Their `search-first` skill enforces a research phase before any implementation.

### The Pattern

```
1. NEED ANALYSIS — What exactly am I trying to build?
   - Core requirement (one sentence)
   - Constraints (performance, size, compatibility)
   - Nice-to-haves vs must-haves

2. PARALLEL SEARCH — Search 4 sources simultaneously:
   - npm/PyPI/crates (package registry)
   - GitHub (popular repos, recent activity)
   - Existing codebase (maybe it's already built)
   - Skills/patterns library (maybe there's a pattern)

3. EVALUATION — For each candidate:
   - Maintenance status (last commit, open issues, bus factor)
   - Bundle size / dependency weight
   - API fit (does it solve the actual need?)
   - Security (known vulnerabilities, audit history)

4. DECISION — One of three outcomes:
   - USE existing package (document why)
   - ADAPT existing code (document what changes)
   - BUILD custom (document why nothing fits)
```

### Why This Matters for DevFlow

Our Coder agent jumps straight to implementation. The Explore/Skimmer phase searches the codebase but doesn't search for external solutions. This leads to:
- Reinventing utilities that npm packages handle better
- Missing well-tested libraries in favor of custom implementations
- Accumulating maintenance burden from custom code

### Where It Fits in DevFlow

The skill would be referenced by the Coder agent and activated by ambient mode for BUILD intent:

1. **Coder agent**: Before implementing a utility function, check if a package exists
2. **Ambient BUILD**: When the user says "add date formatting" or "implement retry logic," the skill prompts research before implementation
3. **Plan phase**: Research results inform architecture decisions

### 1. Create `shared/skills/search-first/SKILL.md`

```yaml
---
name: search-first
description: Research existing solutions before writing custom code. Prevents reinventing the wheel.
allowed-tools: Read, Grep, Glob
---
```

Skill content would define:
- The 4-phase research loop
- Evaluation criteria for packages (maintenance, size, security, API fit)
- Decision framework (USE / ADAPT / BUILD with justification)
- Anti-patterns: "Don't add a dependency for 5 lines of code"
- Scope: Only triggers for utility-level code, not domain-specific business logic

### 2. Ambient Router Integration

Add to the BUILD skill selection matrix:
```
| Intent | Primary Skills | Secondary |
|--------|---------------|-----------|
| BUILD  | test-driven-development, implementation-patterns, search-first | ... |
```

### 3. Plugin Assignment

Add to `devflow-core-skills` plugin (it's a universal pattern, not language-specific).

## Scope & Limitations

- **Not for business logic**: Only applies to utility/infrastructure code where packages commonly exist
- **Not blocking**: The skill guides research; it doesn't prevent implementation
- **Lightweight**: ~100 lines of skill content, no agent changes needed
- **Additive**: Loads alongside existing BUILD skills, doesn't replace them

## Effort & Impact

- **Effort**: Small (one skill file + ambient router update + plugin manifest)
- **Impact**: Medium — prevents reinventing the wheel, especially in unfamiliar domains
- **Risk**: None — advisory skill, no enforcement mechanism

### LINKED ISSUE #113: Reviewer Confidence Thresholds: Reduce Noise in Code Review Output
## Summary

Add confidence thresholds to Reviewer agents so only findings with >80% confidence are reported. Reduces noise in `/code-review` output and increases signal-to-noise ratio. Inspired by Harness Alpha's confidence-based filtering pattern.

## Motivation

Harness Alpha's code-reviewer and all language-specific reviewers (Python, Go, Kotlin) require **>80% confidence before flagging issues**. Their philosophy: "Better to miss something than flood with false positives."

### The Problem

Without confidence thresholds, reviewers report everything they notice — including:
- Style preferences that aren't clearly wrong
- Potential issues that depend on context the reviewer can't see
- Edge cases that are handled elsewhere in the codebase
- "Could be improved" suggestions that distract from real issues

This creates review fatigue. When 40% of findings are noise, users start ignoring all findings.

### The Pattern

Each review finding must include a confidence assessment:

```markdown
### Finding: Missing error boundary around API call
- **Severity**: HIGH
- **Confidence**: 90%
- **File**: src/api/client.ts:45
- **Issue**: Uncaught promise rejection could crash the app
- **Fix**: Wrap in try/catch with error reporting

### Finding: Variable name could be more descriptive
- **Severity**: LOW
- **Confidence**: 55%  ← FILTERED (below 80% threshold)
```

Only findings with confidence ≥ 80% appear in the review output. Lower-confidence findings are either dropped or collected in a separate "suggestions" section.

### Why This Matters for DevFlow

Our `/code-review` spawns 7-11 Reviewer agents in parallel. Each reports all findings regardless of confidence. The Synthesizer deduplicates and merges, but doesn't filter by confidence — noise from individual reviewers propagates to the final report.

Adding confidence thresholds would:
1. Reduce review report length by ~30-40% (estimated noise ratio)
2. Increase user trust in review output
3. Focus attention on real issues

### 1. Update Reviewer Agent Prompt

Add confidence assessment requirement to `shared/agents/reviewer.md`:

```markdown
## Finding Format

For each issue found, assess your confidence (0-100%):
- **90-100%**: Certain — clear bug, security vulnerability, or standards violation
- **80-89%**: High — likely issue based on context and patterns
- **60-79%**: Medium — possible issue, depends on context not visible
- **Below 60%**: Low — subjective preference or uncertain

**Only report findings with confidence ≥ 80%.** 

Collect lower-confidence observations in a separate "Suggestions" section (max 3 items).
```

### 2. Structured Output

```markdown
## Critical & High Confidence Findings
[Only ≥80% confidence findings, severity-ordered]

## Suggestions (Lower Confidence)
[Max 3 items, clearly labeled as suggestions not findings]
```

### 3. Synthesizer Integration

Update Synthesizer to:
- Respect confidence levels during deduplication
- Boost confidence when multiple reviewers flag the same issue
- Maintain the ≥80% threshold in the final report

## Effort & Impact

- **Effort**: Small (agent prompt update + synthesizer tweak)
- **Impact**: Medium — cleaner reviews, higher user trust, less review fatigue
- **Risk**: Low — worst case is missing a marginal issue that a user would have ignored anyway

## Cross-Reference

- Enhances `/code-review` output quality
- Complements #107 item 9 (De-Sloppify Categories) — both aim to increase signal-to-noise
- Independent of #98 (PreToolUse enforcement) and #110 (PostToolUse quality hooks)

### PR DESCRIPTION
**Pull request by dean0x**
## Summary

Implements the three independent Wave 1 features from the post-v1.0 roadmap (#137):

- **Search-first skill** (#111) — New skill enforcing a 4-phase research loop (Need Analysis → Search → Evaluate → Decide) before writing custom utility code. Added to core-skills plugin and ambient-router BUILD primary skills. 31 skills total (was 30).

- **Reviewer confidence thresholds** (#113) — Each review finding now includes a visible confidence score (0-100%). Only ≥80% findings appear in main report sections; 60-79% go to a capped Suggestions section. Adds consolidation rules (group similar issues, skip stylistic preferences, skip unchanged code unless CRITICAL). Fixes synthesizer review glob bug that matched zero reviewer files.

- **Version manifest** (#91) — Tracks installed version, plugins, scope, and feature flags in `manifest.json`. Enables upgrade detection during `devflow init` and shows install status in `devflow list`. Partial installs merge plugins instead of overwriting. 17 new tests.

Closes #111, closes #113, closes #91

## Test plan

- [x] `npm run build` passes — 31 skills, 17 plugins
- [x] `npm test` passes — 208 tests (was 191, +17 manifest tests)
- [x] `search-first` skill distributed to `plugins/devflow-core-skills/skills/`
- [x] Zero regressions on existing tests
- [ ] Manual: run `node dist/cli.js list` to verify install status display
- [ ] Manual: run `/code-review` on a test branch to verify confidence percentages and suggestions section

### CODE REVIEW

File: src/cli/utils/manifest.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+/**
+ * Read and parse the manifest file. Returns null if missing or corrupt.
+ */
+export async function readManifest(devflowDir: string): Promise<ManifestData | null> {
+  const manifestPath = path.join(devflowDir, 'manifest.json');
+  try {
+    const content = await fs.readFile(manifestPath, 'utf-8');
+    const data = JSON.parse(content) as ManifestData;
```
Comment (dean0x) [inline src/cli/utils/manifest.ts]: ## Manifest Validation Gap (BLOCKING)

The `readManifest` function validates only 3 fields (`version`, `plugins`, `scope`) but casts the full object via `as ManifestData`. Missing validation for `features` object, `installedAt`, and `updatedAt`.

**Risk**: Partial/malformed manifests pass validation but cause runtime crashes in `list.ts:41` when accessing `manifest.features.teams`.

**Fix**: Add full field validation:

```typescript
if (
  \!data.version ||
  \!Array.isArray(data.plugins) ||
  \!data.scope ||
  typeof data.features \!== "object" ||
  data.features === null ||
  typeof data.features.teams \!== "boolean" ||
  typeof data.features.ambient \!== "boolean" ||
  typeof data.features.memory \!== "boolean" ||
  \!data.installedAt ||
  \!data.updatedAt
) {
  return null;
}
```

Or use Zod schema (project convention per CLAUDE.md).

**Confidence**: 85% | **Consolidated**: Reported in 5 reviews (security, architecture, consistency, tests, regression)

_Claude Code Review_

**Comment by dean0x** [comment]
## Code Review Summary — Blocking & Should-Fix Issues

### Blocking Issues (2 inline comments created above)

✅ **Manifest Validation Gap** (`src/cli/utils/manifest.ts:27-30`) - FIXED via inline comment
- Validates only 3 fields but casts full object; can crash when accessing undefined `features` properties
- Confidence: 85% | Consolidated from 5 reviews

✅ **Missing Error Handling** (`src/cli/commands/init.ts:624`) - FIXED via inline comment  
- `writeManifest()` not wrapped in try/catch; write failure crashes after successful install
- Confidence: 82% | Consolidated from 2 reviews

### Documentation Issues (HIGH - Not in PR diff)

⚠️ **README.md outdated** — These files are not modified in this PR but need updates:
- **Line 27**: Says "30 quality skills" but CLAUDE.md updated to 31
  - Fix: Change to "31 quality skills" and note "9 auto-activating core" (was 8)
- **Lines 109-122**: Auto-activating skills table missing `search-first` entry
  - The skill was added to `devflow-core-skills` plugin.json but README table was not updated
  - Confidence: 95% | Category: HIGH

**Action**: Please update README.md in a follow-up commit to prevent documentation-reality drift.

---

### Should-Fix Issues

#### 📊 Test Coverage Gaps (HIGH Risk)

**Missing edge case tests** (`tests/manifest.test.ts` / `src/cli/utils/manifest.ts`)
- `readManifest` missing test for partial manifest (missing `features`/`installedAt`/`updatedAt`)
- `compareSemver` v-prefix path untested (regex has `v?` but no test exercises it)
- Suggested fixes in review reports (tests.md)

**Zero test coverage for integration logic** (`src/cli/commands/list.ts`, `src/cli/commands/init.ts`)
- `list.ts` manifest reading + scope resolution + feature string logic untested
- `init.ts` conditional plugin merge logic untested (especially risky: `existingManifest && options.plugin`)
- Recommended: Extract pure functions (`formatFeatures`, `resolvePluginList`) for testability

Confidence: 85-88% | Category: HIGH

#### ⚡ Performance Opportunity

**Sequential I/O in `list` command** (`src/cli/commands/list.ts:17-21`)
- `getGitRoot()` and `readManifest(userDevflowDir)` run sequentially but are independent
- Could parallelize with `Promise.all()` for faster execution on slow filesystems
- Confidence: 82% | Category: MEDIUM

#### 🎯 Complexity Reduction

**Extract `formatInstallStatus()` helper** (`src/cli/commands/list.ts:54`)
- Manifest display logic (127-char ternary) could move to dedicated function
- Would reduce action handler nesting and improve readability
- Confidence: 80% | Category: MEDIUM

**Extract manifest helpers from `init.ts`** (`src/cli/commands/init.ts:296-305,611-624`)
- `logUpgradeStatus()` and `saveManifest()` could separate concerns
- Keeps action handler from growing further (already 515 lines with 15+ responsibilities)
- Confidence: 85% | Category: MEDIUM

---

### Summary by Severity

| Category | Count | Details |
|----------|-------|---------|
| **Blocking** | 2 | Manifest validation, error handling (inlined above) |
| **HIGH** | 3 | README drift (2 items) + test coverage gaps |
| **MEDIUM** | 4 | Performance, complexity extraction (2 files) |

_Claude Code Review Assistant_
